In [0]:
%python
from csv import reader
from pyspark.sql import Row 
from pyspark.sql import SparkSession
from pyspark.sql.types import *
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
import warnings

import os
os.environ["PYSPARK_PYTHON"] = "python3"


In [0]:
%python
data_path = "dbfs:/laioffer/spark_hw1/data/sf_03_18.csv"
# use this file name later

### Solove  big data issues via Spark
approach 1: use RDD (not recommend)  
approach 2: use Dataframe, register the RDD to a dataframe (recommend for DE)  
approach 3: use SQL (recomend for data analysis or DS， 基础比较差的同学)  
***note***: you only need to choose one of approaches as introduced above

In [0]:
%python

from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("crime analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

df_opt1 = spark.read.format("csv").option("header", "true").load(data_path)
display(df_opt1)
df_opt1.createOrReplaceTempView("sf_crime")






PdId IncidntNum Incident Code Category Descript DayOfWeek Date Time PdDistrict Resolution Address X Y location SF Find Neighborhoods 2 2 Current Police Districts 2 2 Current Supervisor Districts 2 2 Analysis Neighborhoods 2 2 DELETE - Fire Prevention Districts 2 2 DELETE - Police Districts 2 2 DELETE - Supervisor Districts 2 2 DELETE - Zip Codes 2 2 DELETE - Neighborhoods 2 2 DELETE - 2017 Fix It Zones 2 2 Civic Center Harm Reduction Project Boundary 2 2 Fix It Zones as of 2017-11-06 2 2 DELETE - HSOC Zones 2 2 Fix It Zones as of 2018-02-07 2 2 CBD, BID and GBD Boundaries as of 2017 2 2 Areas of Vulnerability, 2016 2 2 Central Market/Tenderloin Boundary 2 2 Central Market/Tenderloin Boundary Polygon - Updated 2 2 HSOC Zones as of 2018-06-05 2 2 OWED Public Spaces 2 2 Neighborhoods 2 3114751606302 031147516 06302 LARCENY/THEFT PETTY THEFT FROM A BUILDING Sunday 09/28/2003 10:00 SOUTHERN NONE 1ST ST / BRANNAN ST -120.5 90 POINT (-120.50000000000001 90) null null null null null null null null null null null null null null null null null null null null null 5069701104134 050697011 04134 ASSAULT BATTERY Wednesday 06/22/2005 12:20 NORTHERN NONE 1200 Block of EDDY ST -122.428223303176 37.7818959488603 POINT (-122.42822330317601 37.7818959488603) 97 4 11 39 7 9 11 29490 41 null null null null null null 2 null null null null 97 6074729204104 060747292 04104 ASSAULT ASSAULT Saturday 07/15/2006 00:55 CENTRAL NONE 1600 Block of POWELL ST -122.410672425337 37.799788690123 POINT (-122.41067242533701 37.799788690123) 106 6 3 6 3 1 10 308 4 null null null null null null 2 null null null null 106 7103536315201 071035363 15201 ASSAULT STALKING Tuesday 09/25/2007 00:01 TARAVAL NONE 400 Block of ULLOA ST -122.458226300605 37.7413616001449 POINT (-122.458226300605 37.7413616001449) 49 10 8 41 1 8 4 59 40 null null null null null null 1 null null null null 49 11082415274000 110824152 74000 MISSING PERSON MISSING ADULT Saturday 09/24/2011 11:00 TARAVAL LOCATED 3200 Block of SAN JOSE AV -122.459172646607 37.7082001648459 POINT (-122.459172646607 37.7082001648459) null null null null null null null 28588 null null null null null null null null null null null null null 4037801104134 040378011 04134 ASSAULT BATTERY Friday 12/12/2003 12:00 SOUTHERN NONE 500 Block of I-80 -122.386667033903 37.7898821569191 POINT (-122.38666703390301 37.7898821569191) null null null null null null null 28856 null null null null null null null null null null null null null 4147669007025 041476690 07025 VEHICLE THEFT STOLEN TRUCK Thursday 12/30/2004 19:00 BAYVIEW NONE 100 Block of KAREN CT -122.407531316146 37.723291086507 POINT (-122.40753131614602 37.723291086507) 91 2 2 25 10 3 7 309 28 null null null null null null 1 null null null null 91 16010127305073 160101273 05073 BURGLARY BURGLARY, UNLAWFUL ENTRY Wednesday 02/03/2016 20:30 MISSION ARREST, BOOKED 2300 Block of 16TH ST -122.40952991327822 37.76571843952822 POINT (-122.40952991327822 37.76571843952822) 53 3 9 20 8 4 8 28853 19 null null null 3 null null 2 null null 3 null 53 17004924306243 170049243 06243 LARCENY/THEFT PETTY THEFT FROM LOCKED AUTO Wednesday 01/18/2017 18:20 TARAVAL NONE NORIEGA ST / SUNSET BL -122.49522504431104 37.75341185338174 POINT (-122.49522504431104 37.75341185338174) 39 10 7 35 1 8 3 56 35 null null null null null null 1 null null null null 39 16065828006244 160658280 06244 LARCENY/THEFT GRAND THEFT FROM LOCKED AUTO Sunday 08/14/2016 21:00 TARAVAL NONE 3200 Block of 20TH AV -122.47564746078615 37.72852806274646 POINT (-122.47564746078615 37.72852806274647) 41 10 8 16 1 8 4 64 14 null null null null null null 2 null null null null 41 18032260316100 180322603 16100 DRUG/NARCOTIC POSSESSION OF HEROIN Tuesday 05/01/2018 13:27 SOUTHERN ARREST, BOOKED 600 Block of STEVENSON ST -122.41265189435406 37.77951316688882 POINT (-122.41265189435406 37.77951316688882) 32 5 10 34 8 2 9 28853 34 3 1 3 1 3 7 2 1 1 1 null 32 17612518006244 176125180 06244 LARCENY/THEFT GRAND THEFT FROM LOCKED AUTO Tuesday 05/16

#### Q1 question (OLAP): 
#####Write a Spark program that counts the number of crimes for different category.

Below are some example codes to demonstrate the way to use Spark RDD, DF, and SQL to work with big data. You can follow this example to finish other questions.

In [0]:
%python
q1_result = df_opt1.groupBy('category').count().orderBy('count', ascending=False)
display(q1_result)

category count LARCENY/THEFT 477975 OTHER OFFENSES 301874 NON-CRIMINAL 236928 ASSAULT 167042 VEHICLE THEFT 126228 DRUG/NARCOTIC 117821 VANDALISM 114718 WARRANTS 99821 BURGLARY 91067 SUSPICIOUS OCC 79087 ROBBERY 54467 MISSING PERSON 44268 FRAUD 41348 FORGERY/COUNTERFEITING 22995 SECONDARY CODES 22378 WEAPON LAWS 21004 TRESPASS 19194 PROSTITUTION 16501 STOLEN PROPERTY 11450 DISORDERLY CONDUCT 9932 DRUNKENNESS 9760 SEX OFFENSES, FORCIBLE 8747 RECOVERED VEHICLE 8688 DRIVING UNDER THE INFLUENCE 5652 KIDNAPPING 4282 ARSON 3875 EMBEZZLEMENT 2984 LIQUOR LAWS 2840 LOITERING 2402 SUICIDE 1291 BAD CHECKS 924 BRIBERY 796 EXTORTION 729 GAMBLING 343 PORNOGRAPHY/OBSCENE MAT 57 SEX OFFENSES, NON FORCIBLE 43 TREA 14

In [0]:
%sql

SELECT  category, COUNT(*) AS Count FROM sf_crime GROUP BY category ORDER BY Count DESC


category Count LARCENY/THEFT 477975 OTHER OFFENSES 301874 NON-CRIMINAL 236928 ASSAULT 167042 VEHICLE THEFT 126228 DRUG/NARCOTIC 117821 VANDALISM 114718 WARRANTS 99821 BURGLARY 91067 SUSPICIOUS OCC 79087 ROBBERY 54467 MISSING PERSON 44268 FRAUD 41348 FORGERY/COUNTERFEITING 22995 SECONDARY CODES 22378 WEAPON LAWS 21004 TRESPASS 19194 PROSTITUTION 16501 STOLEN PROPERTY 11450 DISORDERLY CONDUCT 9932 DRUNKENNESS 9760 SEX OFFENSES, FORCIBLE 8747 RECOVERED VEHICLE 8688 DRIVING UNDER THE INFLUENCE 5652 KIDNAPPING 4282 ARSON 3875 EMBEZZLEMENT 2984 LIQUOR LAWS 2840 LOITERING 2402 SUICIDE 1291 BAD CHECKS 924 BRIBERY 796 EXTORTION 729 GAMBLING 343 PORNOGRAPHY/OBSCENE MAT 57 SEX OFFENSES, NON FORCIBLE 43 TREA 14

In [0]:
%sql

SELECT  category, COUNT(*) AS Count FROM sf_crime GROUP BY category ORDER BY Count DESC


category Count LARCENY/THEFT 477975 OTHER OFFENSES 301874 NON-CRIMINAL 236928 ASSAULT 167042 VEHICLE THEFT 126228 DRUG/NARCOTIC 117821 VANDALISM 114718 WARRANTS 99821 BURGLARY 91067 SUSPICIOUS OCC 79087 ROBBERY 54467 MISSING PERSON 44268 FRAUD 41348 FORGERY/COUNTERFEITING 22995 SECONDARY CODES 22378 WEAPON LAWS 21004 TRESPASS 19194 PROSTITUTION 16501 STOLEN PROPERTY 11450 DISORDERLY CONDUCT 9932 DRUNKENNESS 9760 SEX OFFENSES, FORCIBLE 8747 RECOVERED VEHICLE 8688 DRIVING UNDER THE INFLUENCE 5652 KIDNAPPING 4282 ARSON 3875 EMBEZZLEMENT 2984 LIQUOR LAWS 2840 LOITERING 2402 SUICIDE 1291 BAD CHECKS 924 BRIBERY 796 EXTORTION 729 GAMBLING 343 PORNOGRAPHY/OBSCENE MAT 57 SEX OFFENSES, NON FORCIBLE 43 TREA 14

Top three crimes include Larceny/Theft, Other offenses and Non-criminal. Problem is that why Non-criminal is a kind of criminal. The brief search gave that activities like any damage caused to property or things of an unintentional or non-malicious nature, and any incident that results in the damage to a building or things and not arson related are included as non-criminal crime. We may discuss with data provider to confirm the meaning of this category.

In [0]:
%python
# important hints: 
## first step: spark df or sql to compute the statisitc result 
## second step: export your result to a pandas dataframe. 
crimeCategory=spark.sql("SELECT  category, COUNT(*) AS Count FROM sf_crime GROUP BY category ORDER BY Count DESC")
crimes_pd_df = crimeCategory.toPandas()
crimes_pd_df.head()
# Spark does not support this function, please refer https://matplotlib.org/ for visuliation. You need to use display to show the figure in the databricks community. 

#display(p)

category 
 Count 
 
 
 
 
 0 
 LARCENY/THEFT 
 477975 
 
 
 1 
 OTHER OFFENSES 
 301874 
 
 
 2 
 NON-CRIMINAL 
 236928 
 
 
 3 
 ASSAULT 
 167042 
 
 
 4 
 VEHICLE THEFT 
 126228

#### Q2 question (OLAP)
Counts the number of crimes for different district, and visualize your results

In [0]:
%python
crimeDistrict = spark.sql("SELECT  PdDistrict, COUNT(*) AS Count FROM sf_crime GROUP BY PdDistrict ORDER BY Count DESC")
display(crimeDistrict)

PdDistrict Count SOUTHERN 390692 MISSION 288985 NORTHERN 266435 CENTRAL 221923 BAYVIEW 205480 TENDERLOIN 186954 INGLESIDE 181092 TARAVAL 155461 PARK 119698 RICHMOND 112804 NA 1

In [0]:
%python
crimeDistrict = spark.sql("SELECT  PdDistrict, COUNT(*) AS Count FROM sf_crime GROUP BY PdDistrict ORDER BY Count DESC")
display(crimeDistrict)

PdDistrict Count SOUTHERN 390692 MISSION 288985 NORTHERN 266435 CENTRAL 221923 BAYVIEW 205480 TENDERLOIN 186954 INGLESIDE 181092 TARAVAL 155461 PARK 119698 RICHMOND 112804 NA 1

The southern district highly overwhelms the other teritory for the number of crime numbers. Keep out of this location when buying houses.

#### Q3 question (OLAP)
Count the number of crimes each "Sunday" at "SF downtown".   
hint 1: SF downtown is defiend  via the range of spatial location. For example, you can use a rectangle to define the SF downtown, or you can define a cicle with center as well. Thus, you need to write your own UDF function to filter data which are located inside certain spatial range. You can follow the example here: https://changhsinlee.com/pyspark-udf/

hint 2: SF downtown 物理范围可以是 rectangle a < x < b  and c < y < d. thus, San Francisco Latitude and longitude coordinates are: 37.773972, -122.431297. X and Y represents each. So we assume SF downtown spacial range: X (-122.4213,-122.4313), Y(37.7540,37.7740). 也可以是中心一个圈，距离小于多少算做downtown

In [0]:
%sql

SELECT Date, DayOfWeek, COUNT(*) AS Count
FROM sf_crime 
WHERE DayOfWeek = 'Sunday'
AND X > -122.4313 
AND X < -122.4213 
AND Y > 37.7540 
AND Y < 37.7740 
GROUP BY Date, DayOfWeek
ORDER BY Count


Date DayOfWeek Count 09/23/2007 Sunday 1 12/25/2011 Sunday 2 07/27/2008 Sunday 2 11/14/2010 Sunday 3 08/14/2005 Sunday 3 05/06/2007 Sunday 3 04/17/2005 Sunday 4 09/12/2010 Sunday 4 03/27/2011 Sunday 4 12/07/2003 Sunday 4 12/14/2008 Sunday 4 01/15/2006 Sunday 4 08/24/2014 Sunday 4 08/21/2011 Sunday 4 12/23/2007 Sunday 5 03/18/2007 Sunday 5 02/19/2012 Sunday 5 05/23/2004 Sunday 5 12/24/2006 Sunday 5 08/22/2010 Sunday 5 03/06/2011 Sunday 5 07/20/2014 Sunday 5 08/10/2003 Sunday 5 10/12/2008 Sunday 5 05/11/2003 Sunday 5 02/02/2014 Sunday 5 04/27/2003 Sunday 5 06/22/2008 Sunday 5 02/16/2014 Sunday 5 07/11/2010 Sunday 5 01/15/2012 Sunday 5 09/18/2005 Sunday 5 05/13/2007 Sunday 5 04/04/2010 Sunday 5 02/06/2011 Sunday 6 08/02/2009 Sunday 6 07/18/2010 Sunday 6 03/20/2011 Sunday 6 02/04/2007 Sunday 6 08/09/2009 Sunday 6 03/13/2016 Sunday 6 12/11/2005 Sunday 6 12/26/2010 Sunday 6 11/30/2003 Sunday 6 05/22/2005 Sunday 6 07/05/2009 Sunday 6 02/27/2011 Sunday 6 09/28/2014 Sunday 6 06/22/2003 Sunday 6 07/19/2015 Sunday 6 08/03/2003 Sunday 6 08/06/2006 Sunday 6 03/27/2005 Sunday 6 03/04/2007 Sunday 6 01/18/2004 Sunday 6 09/13/2015 Sunday 6 01/27/2008 Sunday 6 09/24/2006 Sunday 6 03/12/2006 Sunday 6 01/14/2007 Sunday 6 06/15/2003 Sunday 6 09/21/2014 Sunday 7 03/30/2014 Sunday 7 11/12/2017 Sunday 7 07/12/2009 Sunday 7 01/25/2015 Sunday 7 05/07/2006 Sunday 7 04/10/2011 Sunday 7 04/16/2006 Sunday 7 01/21/2007 Sunday 7 01/02/2011 Sunday 7 12/13/2009 Sunday 7 02/17/2008 Sunday 7 12/11/2011 Sunday 7 07/20/2008 Sunday 7 09/28/2003 Sunday 7 03/05/2006 Sunday 7 05/15/2005 Sunday 7 07/10/2011 Sunday 7 01/06/2008 Sunday 7 05/01/2011 Sunday 7 12/18/2005 Sunday 7 11/22/2009 Sunday 7 09/17/2006 Sunday 7 06/03/2007 Sunday 7 11/03/2013 Sunday 7 08/07/2011 Sunday 7 05/13/2018 Sunday 7 02/13/2005 Sunday 7 10/16/2005 Sunday 7 03/30/2008 Sunday 7 02/08/2009 Sunday 7 12/26/2004 Sunday 7 04/11/2010 Sunday 7 09/30/2007 Sunday 7 04/05/2009 Sunday 7 10/04/2015 Sunday 7 01/11/2015 Sunday 8 03/08/2015 Sunday 8 08/05/2007 Sunday 8 08/09/2015 Sunday 8 10/21/2007 Sunday 8 03/02/2008 Sunday 8 11/28/2010 Sunday 8 12/09/2012 Sunday 8 04/06/2003 Sunday 8 09/11/2005 Sunday 8 07/29/2007 Sunday 8 07/27/2003 Sunday 8 08/31/2003 Sunday 8 10/15/2006 Sunday 8 03/22/2009 Sunday 8 11/06/2005 Sunday 8 07/25/2010 Sunday 8 03/09/2003 Sunday 8 07/22/2007 Sunday 8 11/29/2009 Sunday 8 12/19/2010 Sunday 8 09/03/2006 Sunday 8 03/21/2004 Sunday 8 03/20/2005 Sunday 8 12/02/2007 Sunday 8 02/24/2008 Sunday 8 01/02/2005 Sunday 8 10/09/2011 Sunday 8 04/06/2008 Sunday 8 01/16/2005 Sunday 8 04/27/2008 Sunday 8 03/14/2004 Sunday 8 02/25/2007 Sunday 8 11/28/2004 Sunday 8 12/21/2008 Sunday 8 11/13/2011 Sunday 8 08/19/2007 Sunday 8 12/20/2015 Sunday 8 08/15/2004 Sunday 8 12/27/2015 Sunday 8 02/18/2018 Sunday 8 01/24/2016 Sunday 8 11/02/2014 Sunday 8 12/25/2016 Sunday 8 10/08/2017 Sunday 9 11/30/2008 Sunday 9 09/07/2014 Sunday 9 02/23/2014 Sunday 9 07/10/2016 Sunday 9 06/18/2017 Sunday 9 05/01/2016 Sunday 9 05/17/2015 Sunday 9 03/06/2016 Sunday 9 01/22/2017 Sunday 9 11/09/2014 Sunday 9 02/28/2016 Sunday 9 04/09/2017 Sunday 9 11/30/2014 Sunday 9 12/03/2017 Sunday 9 02/14/2010 Sunday 9 07/04/2010 Sunday 9 07/17/2005 Sunday 9 02/26/2006 Sunday 9 10/29/2006 Sunday 9 08/17/2008 Sunday 9 07/21/2013 Sunday 9 06/04/2006 Sunday 9 01/20/2013 Sunday 9 05/08/2011 Sunday 9 06/05/2005 Sunday 9 09/11/2011 Sunday 9 06/11/2006 Sunday 9 02/10/2008 Sunday 9 08/03/2008 Sunday 9 09/04/2005 Sunday 9 05/10/2009 Sunday 9 04/19/2015 Sunday 9 01/06/2013 Sunday 9 09/02/2007 Sunday 9 10/23/2005 Sunday 9 07/06/2008 Sunday 9 10/30/2005 Sunday 9 09/18/2011 Sunday 9 01/03/2016 Sunday 9 04/01/2018 Sunday 9 09/12/2004 Sunday 9 01/08/2017 Sunday 9 11/13/2016 Sunday 9 05/04/2003 Sunday 9 12/05/2004 Sunday 9 07/15/2007 Sunday 9 11/01/2009 Sunday 9 02/07/2010 Sunday 9 03/23/2003 Sunday 9 12/17/2006 Sunday 9 11/26/2006 Sunday 9 12/13/2015 Sunday 9 05/11/2008 Sunday 9 11/15/2015 Sunday 9 06/20/2004 Sunday 10 08/07/2005 Sunday 10 03/26/2006 Sunday

In [0]:
%sql

SELECT Date, DayOfWeek, COUNT(*) AS Count
FROM sf_crime 
WHERE DayOfWeek = 'Sunday'
AND X > -122.4313 
AND X < -122.4213 
AND Y > 37.7540 
AND Y < 37.7740 
GROUP BY Date, DayOfWeek
ORDER BY Count

Date DayOfWeek Count 09/23/2007 Sunday 1 12/25/2011 Sunday 2 07/27/2008 Sunday 2 11/14/2010 Sunday 3 08/14/2005 Sunday 3 05/06/2007 Sunday 3 04/17/2005 Sunday 4 09/12/2010 Sunday 4 03/27/2011 Sunday 4 12/07/2003 Sunday 4 12/14/2008 Sunday 4 01/15/2006 Sunday 4 08/24/2014 Sunday 4 08/21/2011 Sunday 4 12/23/2007 Sunday 5 03/18/2007 Sunday 5 02/19/2012 Sunday 5 05/23/2004 Sunday 5 12/24/2006 Sunday 5 08/22/2010 Sunday 5 03/06/2011 Sunday 5 07/20/2014 Sunday 5 08/10/2003 Sunday 5 10/12/2008 Sunday 5 05/11/2003 Sunday 5 02/02/2014 Sunday 5 04/27/2003 Sunday 5 06/22/2008 Sunday 5 02/16/2014 Sunday 5 07/11/2010 Sunday 5 01/15/2012 Sunday 5 09/18/2005 Sunday 5 05/13/2007 Sunday 5 04/04/2010 Sunday 5 02/06/2011 Sunday 6 08/02/2009 Sunday 6 07/18/2010 Sunday 6 03/20/2011 Sunday 6 02/04/2007 Sunday 6 08/09/2009 Sunday 6 03/13/2016 Sunday 6 12/11/2005 Sunday 6 12/26/2010 Sunday 6 11/30/2003 Sunday 6 05/22/2005 Sunday 6 07/05/2009 Sunday 6 02/27/2011 Sunday 6 09/28/2014 Sunday 6 06/22/2003 Sunday 6 07/19/2015 Sunday 6 08/03/2003 Sunday 6 08/06/2006 Sunday 6 03/27/2005 Sunday 6 03/04/2007 Sunday 6 01/18/2004 Sunday 6 09/13/2015 Sunday 6 01/27/2008 Sunday 6 09/24/2006 Sunday 6 03/12/2006 Sunday 6 01/14/2007 Sunday 6 06/15/2003 Sunday 6 09/21/2014 Sunday 7 03/30/2014 Sunday 7 11/12/2017 Sunday 7 07/12/2009 Sunday 7 01/25/2015 Sunday 7 05/07/2006 Sunday 7 04/10/2011 Sunday 7 04/16/2006 Sunday 7 01/21/2007 Sunday 7 01/02/2011 Sunday 7 12/13/2009 Sunday 7 02/17/2008 Sunday 7 12/11/2011 Sunday 7 07/20/2008 Sunday 7 09/28/2003 Sunday 7 03/05/2006 Sunday 7 05/15/2005 Sunday 7 07/10/2011 Sunday 7 01/06/2008 Sunday 7 05/01/2011 Sunday 7 12/18/2005 Sunday 7 11/22/2009 Sunday 7 09/17/2006 Sunday 7 06/03/2007 Sunday 7 11/03/2013 Sunday 7 08/07/2011 Sunday 7 05/13/2018 Sunday 7 02/13/2005 Sunday 7 10/16/2005 Sunday 7 03/30/2008 Sunday 7 02/08/2009 Sunday 7 12/26/2004 Sunday 7 04/11/2010 Sunday 7 09/30/2007 Sunday 7 04/05/2009 Sunday 7 10/04/2015 Sunday 7 01/11/2015 Sunday 8 03/08/2015 Sunday 8 08/05/2007 Sunday 8 08/09/2015 Sunday 8 10/21/2007 Sunday 8 03/02/2008 Sunday 8 11/28/2010 Sunday 8 12/09/2012 Sunday 8 04/06/2003 Sunday 8 09/11/2005 Sunday 8 07/29/2007 Sunday 8 07/27/2003 Sunday 8 08/31/2003 Sunday 8 10/15/2006 Sunday 8 03/22/2009 Sunday 8 11/06/2005 Sunday 8 07/25/2010 Sunday 8 03/09/2003 Sunday 8 07/22/2007 Sunday 8 11/29/2009 Sunday 8 12/19/2010 Sunday 8 09/03/2006 Sunday 8 03/21/2004 Sunday 8 03/20/2005 Sunday 8 12/02/2007 Sunday 8 02/24/2008 Sunday 8 01/02/2005 Sunday 8 10/09/2011 Sunday 8 04/06/2008 Sunday 8 01/16/2005 Sunday 8 04/27/2008 Sunday 8 03/14/2004 Sunday 8 02/25/2007 Sunday 8 11/28/2004 Sunday 8 12/21/2008 Sunday 8 11/13/2011 Sunday 8 08/19/2007 Sunday 8 12/20/2015 Sunday 8 08/15/2004 Sunday 8 12/27/2015 Sunday 8 02/18/2018 Sunday 8 01/24/2016 Sunday 8 11/02/2014 Sunday 8 12/25/2016 Sunday 8 10/08/2017 Sunday 9 11/30/2008 Sunday 9 09/07/2014 Sunday 9 02/23/2014 Sunday 9 07/10/2016 Sunday 9 06/18/2017 Sunday 9 05/01/2016 Sunday 9 05/17/2015 Sunday 9 03/06/2016 Sunday 9 01/22/2017 Sunday 9 11/09/2014 Sunday 9 02/28/2016 Sunday 9 04/09/2017 Sunday 9 11/30/2014 Sunday 9 12/03/2017 Sunday 9 02/14/2010 Sunday 9 07/04/2010 Sunday 9 07/17/2005 Sunday 9 02/26/2006 Sunday 9 10/29/2006 Sunday 9 08/17/2008 Sunday 9 07/21/2013 Sunday 9 06/04/2006 Sunday 9 01/20/2013 Sunday 9 05/08/2011 Sunday 9 06/05/2005 Sunday 9 09/11/2011 Sunday 9 06/11/2006 Sunday 9 02/10/2008 Sunday 9 08/03/2008 Sunday 9 09/04/2005 Sunday 9 05/10/2009 Sunday 9 04/19/2015 Sunday 9 01/06/2013 Sunday 9 09/02/2007 Sunday 9 10/23/2005 Sunday 9 07/06/2008 Sunday 9 10/30/2005 Sunday 9 09/18/2011 Sunday 9 01/03/2016 Sunday 9 04/01/2018 Sunday 9 09/12/2004 Sunday 9 01/08/2017 Sunday 9 11/13/2016 Sunday 9 05/04/2003 Sunday 9 12/05/2004 Sunday 9 07/15/2007 Sunday 9 11/01/2009 Sunday 9 02/07/2010 Sunday 9 03/23/2003 Sunday 9 12/17/2006 Sunday 9 11/26/2006 Sunday 9 12/13/2015 Sunday 9 05/11/2008 Sunday 9 11/15/2015 Sunday 9 06/20/2004 Sunday 10 08/07/2005 Sunday 10 03/26/2006 Sunday

The crime amounts in each year did not show a clear trend in either increasing or decreasing. The peak exists in year 2013. The distribution is like a bell distribution.

#### Q4 question (OLAP)
Analysis the number of crime in each month of 2015, 2016, 2017, 2018. Then, give your insights for the output results. What is the business impact for your result?

In [0]:
%sql
SELECT SUBSTRING(Date,1,2) AS Month, SUBSTRING(Date,7,4) AS Year, COUNT(*) AS Count
FROM sf_crime
GROUP BY Year, Month
Having Year in (2015,2016,2017,2018)
Order BY Year, Month


Month Year Count 01 2015 13181 02 2015 11882 03 2015 13463 04 2015 12526 05 2015 13318 06 2015 12853 07 2015 12949 08 2015 13317 09 2015 12476 10 2015 12697 11 2015 11691 12 2015 11106 01 2016 12532 02 2016 11743 03 2016 11950 04 2016 11905 05 2016 12313 06 2016 11727 07 2016 11832 08 2016 12104 09 2016 12092 10 2016 12913 11 2016 12254 12 2016 12629 01 2017 12687 02 2017 11780 03 2017 13250 04 2017 12452 05 2017 12758 06 2017 12186 07 2017 12717 08 2017 12428 09 2017 12204 10 2017 12970 11 2017 11940 12 2017 12115 01 2018 11667 02 2018 9565 03 2018 10354 04 2018 9954 05 2018 3519

In [0]:
%sql
SELECT SUBSTRING(Date,1,2) AS Month, SUBSTRING(Date,7,4) AS Year, COUNT(*) AS Count
FROM sf_crime
GROUP BY Year, Month
Having Year in (2015,2016,2017,2018)
Order BY Year, Month
 

Month Year Count 01 2015 13181 02 2015 11882 03 2015 13463 04 2015 12526 05 2015 13318 06 2015 12853 07 2015 12949 08 2015 13317 09 2015 12476 10 2015 12697 11 2015 11691 12 2015 11106 01 2016 12532 02 2016 11743 03 2016 11950 04 2016 11905 05 2016 12313 06 2016 11727 07 2016 11832 08 2016 12104 09 2016 12092 10 2016 12913 11 2016 12254 12 2016 12629 01 2017 12687 02 2017 11780 03 2017 13250 04 2017 12452 05 2017 12758 06 2017 12186 07 2017 12717 08 2017 12428 09 2017 12204 10 2017 12970 11 2017 11940 12 2017 12115 01 2018 11667 02 2018 9565 03 2018 10354 04 2018 9954 05 2018 3519

The amount of crime numbers showed a decreasing trend in these four years. At May 2018, the crime numbers reached the lowest value 3519. It seems that at odd month(like Jan, Mar, May), except 2018, the crime number increased. We may invest insurance company stocks at even month (like Feb, April .etc) because they won't be charged a lot for loss in crimes. On the other hand, we may increase investment in security company stocks at odd month because people may be raised attention for personal and property security.

#### Q5 question (OLAP)
Analysis the number of crime with respsect to the hour in certian day like 2015/12/15, 2016/12/15, 2017/12/15. Then, give your travel suggestion to visit SF.

In [0]:
%sql
SELECT substring(Time,1,2) AS Hour, COUNT(*) AS Count
FROM sf_crime
WHERE Date = '12/15/2017'
GROUP BY Hour
Order BY Hour

Hour Count 00 30 01 12 02 6 03 4 04 9 05 2 06 6 07 11 08 26 09 14 10 27 11 26 12 23 13 10 14 11 15 26 16 30 17 27 18 28 19 29 20 17 21 20 22 36 23 28

In [0]:
%sql
SELECT substring(Time,1,2) AS Hour, COUNT(*) AS Count
FROM sf_crime
WHERE Date = '12/15/2017'
GROUP BY Hour
Order BY Hour

Hour Count 00 30 01 12 02 6 03 4 04 9 05 2 06 6 07 11 08 26 09 14 10 27 11 26 12 23 13 10 14 11 15 26 16 30 17 27 18 28 19 29 20 17 21 20 22 36 23 28

It is better to travel at morning in SF on 12/15/2017.

#### Q6 question (OLAP)
(1) Step1: Find out the top-3 danger disrict  
(2) Step2: find out the crime event w.r.t category and time (hour) from the result of step 1  
(3) give your advice to distribute the police based on your analysis results.

In [0]:
%sql
SELECT PdDistrict, COUNT(*) AS Count
FROM sf_crime
GROUP BY PdDistrict
ORDER BY Count DESC

PdDistrict Count SOUTHERN 390692 MISSION 288985 NORTHERN 266435 CENTRAL 221923 BAYVIEW 205480 TENDERLOIN 186954 INGLESIDE 181092 TARAVAL 155461 PARK 119698 RICHMOND 112804 NA 1

In [0]:
%sql
SELECT PdDistrict, COUNT(*) AS Count
FROM sf_crime
GROUP BY PdDistrict
ORDER BY Count DESC

PdDistrict Count SOUTHERN 390692 MISSION 288985 NORTHERN 266435 CENTRAL 221923 BAYVIEW 205480 TENDERLOIN 186954 INGLESIDE 181092 TARAVAL 155461 PARK 119698 RICHMOND 112804 NA 1

In [0]:
%sql
SELECT Category, Substring(Time,1,2) AS Hour, COUNT(*) AS Count
FROM sf_crime
WHERE PdDistrict = 'SOUTHERN'
GROUP BY Category, Hour
ORDER BY Hour

Category Hour Count VANDALISM 00 1041 EXTORTION 00 15 KIDNAPPING 00 41 ARSON 00 31 LIQUOR LAWS 00 15 VEHICLE THEFT 00 496 WARRANTS 00 746 SECONDARY CODES 00 183 LOITERING 00 8 DRIVING UNDER THE INFLUENCE 00 85 GAMBLING 00 3 SEX OFFENSES, NON FORCIBLE 00 1 LARCENY/THEFT 00 4156 SUSPICIOUS OCC 00 904 BAD CHECKS 00 35 BRIBERY 00 8 BURGLARY 00 486 SUICIDE 00 3 RECOVERED VEHICLE 00 55 EMBEZZLEMENT 00 131 ASSAULT 00 1505 DISORDERLY CONDUCT 00 37 WEAPON LAWS 00 165 FRAUD 00 1172 MISSING PERSON 00 400 OTHER OFFENSES 00 2900 STOLEN PROPERTY 00 122 DRUNKENNESS 00 170 NON-CRIMINAL 00 2408 ROBBERY 00 488 SEX OFFENSES, FORCIBLE 00 202 DRUG/NARCOTIC 00 640 FORGERY/COUNTERFEITING 00 858 TRESPASS 00 100 PROSTITUTION 00 31 PROSTITUTION 01 15 WEAPON LAWS 01 113 MISSING PERSON 01 146 DRUG/NARCOTIC 01 387 OTHER OFFENSES 01 1398 LARCENY/THEFT 01 2576 DRUNKENNESS 01 158 VANDALISM 01 703 SUSPICIOUS OCC 01 399 FRAUD 01 144 SEX OFFENSES, FORCIBLE 01 92 SUICIDE 01 5 EMBEZZLEMENT 01 3 ASSAULT 01 1379 LOITERING 01 3 BAD CHECKS 01 1 DRIVING UNDER THE INFLUENCE 01 101 EXTORTION 01 1 RECOVERED VEHICLE 01 49 LIQUOR LAWS 01 10 SECONDARY CODES 01 109 KIDNAPPING 01 40 ROBBERY 01 518 VEHICLE THEFT 01 308 WARRANTS 01 466 BURGLARY 01 359 FORGERY/COUNTERFEITING 01 49 BRIBERY 01 3 ARSON 01 27 TRESPASS 01 64 STOLEN PROPERTY 01 85 NON-CRIMINAL 01 1240 DISORDERLY CONDUCT 01 24 PROSTITUTION 02 10 EXTORTION 02 1 ROBBERY 02 556 TRESPASS 02 67 BURGLARY 02 399 KIDNAPPING 02 36 RECOVERED VEHICLE 02 26 OTHER OFFENSES 02 1061 SECONDARY CODES 02 111 BRIBERY 02 3 FRAUD 02 108 LARCENY/THEFT 02 1778 EMBEZZLEMENT 02 2 ASSAULT 02 1349 SEX OFFENSES, FORCIBLE 02 99 DRIVING UNDER THE INFLUENCE 02 84 ARSON 02 25 FORGERY/COUNTERFEITING 02 26 SUICIDE 02 5 MISSING PERSON 02 116 WEAPON LAWS 02 99 DRUNKENNESS 02 127 SUSPICIOUS OCC 02 364 DISORDERLY CONDUCT 02 25 NON-CRIMINAL 02 986 VEHICLE THEFT 02 304 WARRANTS 02 356 VANDALISM 02 648 STOLEN PROPERTY 02 46 LOITERING 02 2 DRUG/NARCOTIC 02 277 LIQUOR LAWS 02 10 BAD CHECKS 02 1 KIDNAPPING 03 25 DRIVING UNDER THE INFLUENCE 03 41 NON-CRIMINAL 03 614 FORGERY/COUNTERFEITING 03 27 WARRANTS 03 272 LARCENY/THEFT 03 950 ASSAULT 03 619 VEHICLE THEFT 03 142 VANDALISM 03 469 ROBBERY 03 325 SECONDARY CODES 03 65 DRUNKENNESS 03 44 DRUG/NARCOTIC 03 202 MISSING PERSON 03 56 TRESPASS 03 45 WEAPON LAWS 03 58 STOLEN PROPERTY 03 37 SUSPICIOUS OCC 03 243 BURGLARY 03 417 SEX OFFENSES, FORCIBLE 03 76 EMBEZZLEMENT 03 5 ARSON 03 21 DISORDERLY CONDUCT 03 18 RECOVERED VEHICLE 03 34 LOITERING 03 3 PORNOGRAPHY/OBSCENE MAT 03 1 EXTORTION 03 2 BRIBERY 03 1 LIQUOR LAWS 03 3 OTHER OFFENSES 03 701 PROSTITUTION 03 10 SUICIDE 03 2 FRAUD 03 56 NON-CRIMINAL 04 475 VEHICLE THEFT 04 146 SUSPICIOUS OCC 04 155 DISORDERLY CONDUCT 04 9 TRESPASS 04 39 BRIBERY 04 2 FORGERY/COUNTERFEITING 04 11 DRUNKENNESS 04 22 DRIVING UNDER THE INFLUENCE 04 12 STOLEN PROPERTY 04 31 ARSON 04 20 SECONDARY CODES 04 41 WARRANTS 04 194 MISSING PERSON 04 40 ROBBERY 04 225 OTHER OFFENSES 04 557 FRAUD 04 26 BURGLARY 04 359 KIDNAPPING 04 9 LARCENY/THEFT 04 627 VANDALISM 04 327 DRUG/NARCOTIC 04 131 WEAPON LAWS 04 45 PROSTITUTION 04 5 LIQUOR LAWS 04 4 BAD CHECKS 04 2 LOITERING 04 2 EMBEZZLEMENT 04 3 EXTORTION 04 1 ASSAULT 04 366 SEX OFFENSES, FORCIBLE 04 29 RECOVERED VEHICLE 04 20 ASSAULT 05 304 WEAPON LAWS 05 25 VANDALISM 05 269 DRIVING UNDER THE INFLUENCE 05 7 ROBBERY 05 155 STOLEN PROPERTY 05 28 MISSING PERSON 05 48 FORGERY/COUNTERFEITING 05 9 WARRANTS 05 131 BURGLARY 05 327 ARSON 05 20 OTHER OFFENSES 05 384 LARCENY/THEFT 05 644 DISORDERLY CONDUCT 05 37 NON-CRIMINAL 05 447 RECOVERED VEHICLE 05 14 DRUG/NARCOTIC 05 66 SUSPICIOUS OCC 05 126 EMBEZZLEMENT 05 2 FRAUD 05 32 SEX OFFENSES, FORCIBLE 05 33 TRESPASS 05 69 SECONDARY CODES 05 44 VEHICLE THEFT 05 128 LOITERING 05 7 DRUNKENNESS 05 5 PROSTITUTION 05 1 KIDNAPPING 05 9 BAD CHECKS 05 1 SUICIDE 05 4 EXTORTION 05 1 NON-CRIMINAL 06 803 MISSING PERSON 06 92 OTHER OFFENSES 06 643 FORGERY/COUNTERFEITING 06 36 DRUG/NARCOTIC 06 260 VANDALISM 06 351 EMBEZZLEMENT 06 3 SEX OF

In [0]:
%sql
SELECT Category, Substring(Time,1,2) AS Hour, COUNT(*) AS Count
FROM sf_crime
WHERE PdDistrict = 'SOUTHERN'
GROUP BY Category, Hour
ORDER BY Hour

Category Hour Count VANDALISM 00 1041 EXTORTION 00 15 KIDNAPPING 00 41 ARSON 00 31 LIQUOR LAWS 00 15 VEHICLE THEFT 00 496 WARRANTS 00 746 SECONDARY CODES 00 183 LOITERING 00 8 DRIVING UNDER THE INFLUENCE 00 85 GAMBLING 00 3 SEX OFFENSES, NON FORCIBLE 00 1 LARCENY/THEFT 00 4156 SUSPICIOUS OCC 00 904 BAD CHECKS 00 35 BRIBERY 00 8 BURGLARY 00 486 SUICIDE 00 3 RECOVERED VEHICLE 00 55 EMBEZZLEMENT 00 131 ASSAULT 00 1505 DISORDERLY CONDUCT 00 37 WEAPON LAWS 00 165 FRAUD 00 1172 MISSING PERSON 00 400 OTHER OFFENSES 00 2900 STOLEN PROPERTY 00 122 DRUNKENNESS 00 170 NON-CRIMINAL 00 2408 ROBBERY 00 488 SEX OFFENSES, FORCIBLE 00 202 DRUG/NARCOTIC 00 640 FORGERY/COUNTERFEITING 00 858 TRESPASS 00 100 PROSTITUTION 00 31 PROSTITUTION 01 15 WEAPON LAWS 01 113 MISSING PERSON 01 146 DRUG/NARCOTIC 01 387 OTHER OFFENSES 01 1398 LARCENY/THEFT 01 2576 DRUNKENNESS 01 158 VANDALISM 01 703 SUSPICIOUS OCC 01 399 FRAUD 01 144 SEX OFFENSES, FORCIBLE 01 92 SUICIDE 01 5 EMBEZZLEMENT 01 3 ASSAULT 01 1379 LOITERING 01 3 BAD CHECKS 01 1 DRIVING UNDER THE INFLUENCE 01 101 EXTORTION 01 1 RECOVERED VEHICLE 01 49 LIQUOR LAWS 01 10 SECONDARY CODES 01 109 KIDNAPPING 01 40 ROBBERY 01 518 VEHICLE THEFT 01 308 WARRANTS 01 466 BURGLARY 01 359 FORGERY/COUNTERFEITING 01 49 BRIBERY 01 3 ARSON 01 27 TRESPASS 01 64 STOLEN PROPERTY 01 85 NON-CRIMINAL 01 1240 DISORDERLY CONDUCT 01 24 PROSTITUTION 02 10 EXTORTION 02 1 ROBBERY 02 556 TRESPASS 02 67 BURGLARY 02 399 KIDNAPPING 02 36 RECOVERED VEHICLE 02 26 OTHER OFFENSES 02 1061 SECONDARY CODES 02 111 BRIBERY 02 3 FRAUD 02 108 LARCENY/THEFT 02 1778 EMBEZZLEMENT 02 2 ASSAULT 02 1349 SEX OFFENSES, FORCIBLE 02 99 DRIVING UNDER THE INFLUENCE 02 84 ARSON 02 25 FORGERY/COUNTERFEITING 02 26 SUICIDE 02 5 MISSING PERSON 02 116 WEAPON LAWS 02 99 DRUNKENNESS 02 127 SUSPICIOUS OCC 02 364 DISORDERLY CONDUCT 02 25 NON-CRIMINAL 02 986 VEHICLE THEFT 02 304 WARRANTS 02 356 VANDALISM 02 648 STOLEN PROPERTY 02 46 LOITERING 02 2 DRUG/NARCOTIC 02 277 LIQUOR LAWS 02 10 BAD CHECKS 02 1 KIDNAPPING 03 25 DRIVING UNDER THE INFLUENCE 03 41 NON-CRIMINAL 03 614 FORGERY/COUNTERFEITING 03 27 WARRANTS 03 272 LARCENY/THEFT 03 950 ASSAULT 03 619 VEHICLE THEFT 03 142 VANDALISM 03 469 ROBBERY 03 325 SECONDARY CODES 03 65 DRUNKENNESS 03 44 DRUG/NARCOTIC 03 202 MISSING PERSON 03 56 TRESPASS 03 45 WEAPON LAWS 03 58 STOLEN PROPERTY 03 37 SUSPICIOUS OCC 03 243 BURGLARY 03 417 SEX OFFENSES, FORCIBLE 03 76 EMBEZZLEMENT 03 5 ARSON 03 21 DISORDERLY CONDUCT 03 18 RECOVERED VEHICLE 03 34 LOITERING 03 3 PORNOGRAPHY/OBSCENE MAT 03 1 EXTORTION 03 2 BRIBERY 03 1 LIQUOR LAWS 03 3 OTHER OFFENSES 03 701 PROSTITUTION 03 10 SUICIDE 03 2 FRAUD 03 56 NON-CRIMINAL 04 475 VEHICLE THEFT 04 146 SUSPICIOUS OCC 04 155 DISORDERLY CONDUCT 04 9 TRESPASS 04 39 BRIBERY 04 2 FORGERY/COUNTERFEITING 04 11 DRUNKENNESS 04 22 DRIVING UNDER THE INFLUENCE 04 12 STOLEN PROPERTY 04 31 ARSON 04 20 SECONDARY CODES 04 41 WARRANTS 04 194 MISSING PERSON 04 40 ROBBERY 04 225 OTHER OFFENSES 04 557 FRAUD 04 26 BURGLARY 04 359 KIDNAPPING 04 9 LARCENY/THEFT 04 627 VANDALISM 04 327 DRUG/NARCOTIC 04 131 WEAPON LAWS 04 45 PROSTITUTION 04 5 LIQUOR LAWS 04 4 BAD CHECKS 04 2 LOITERING 04 2 EMBEZZLEMENT 04 3 EXTORTION 04 1 ASSAULT 04 366 SEX OFFENSES, FORCIBLE 04 29 RECOVERED VEHICLE 04 20 ASSAULT 05 304 WEAPON LAWS 05 25 VANDALISM 05 269 DRIVING UNDER THE INFLUENCE 05 7 ROBBERY 05 155 STOLEN PROPERTY 05 28 MISSING PERSON 05 48 FORGERY/COUNTERFEITING 05 9 WARRANTS 05 131 BURGLARY 05 327 ARSON 05 20 OTHER OFFENSES 05 384 LARCENY/THEFT 05 644 DISORDERLY CONDUCT 05 37 NON-CRIMINAL 05 447 RECOVERED VEHICLE 05 14 DRUG/NARCOTIC 05 66 SUSPICIOUS OCC 05 126 EMBEZZLEMENT 05 2 FRAUD 05 32 SEX OFFENSES, FORCIBLE 05 33 TRESPASS 05 69 SECONDARY CODES 05 44 VEHICLE THEFT 05 128 LOITERING 05 7 DRUNKENNESS 05 5 PROSTITUTION 05 1 KIDNAPPING 05 9 BAD CHECKS 05 1 SUICIDE 05 4 EXTORTION 05 1 NON-CRIMINAL 06 803 MISSING PERSON 06 92 OTHER OFFENSES 06 643 FORGERY/COUNTERFEITING 06 36 DRUG/NARCOTIC 06 260 VANDALISM 06 351 EMBEZZLEMENT 06 3 SEX OF

In [0]:
%sql
SELECT Category, Substring(Time,1,2) AS Hour, COUNT(*) AS Count
FROM sf_crime
WHERE PdDistrict ='MISSION'
GROUP BY Category, Hour
ORDER BY Hour

Category Hour Count VANDALISM 00 856 PROSTITUTION 00 706 ROBBERY 00 547 DRUG/NARCOTIC 00 640 BAD CHECKS 00 13 OTHER OFFENSES 00 2703 TRESPASS 00 98 ARSON 00 35 LIQUOR LAWS 00 36 KIDNAPPING 00 40 SECONDARY CODES 00 194 BRIBERY 00 9 VEHICLE THEFT 00 614 WARRANTS 00 666 SEX OFFENSES, NON FORCIBLE 00 3 EXTORTION 00 9 DRIVING UNDER THE INFLUENCE 00 115 GAMBLING 00 6 SUSPICIOUS OCC 00 670 LOITERING 00 23 FORGERY/COUNTERFEITING 00 358 BURGLARY 00 481 WEAPON LAWS 00 188 FRAUD 00 611 RECOVERED VEHICLE 00 40 SUICIDE 00 7 LARCENY/THEFT 00 2553 DISORDERLY CONDUCT 00 188 MISSING PERSON 00 215 ASSAULT 00 1570 DRUNKENNESS 00 155 NON-CRIMINAL 00 1528 EMBEZZLEMENT 00 64 STOLEN PROPERTY 00 71 SEX OFFENSES, FORCIBLE 00 156 TRESPASS 01 71 DISORDERLY CONDUCT 01 129 PROSTITUTION 01 459 WEAPON LAWS 01 157 DRUG/NARCOTIC 01 460 MISSING PERSON 01 114 LARCENY/THEFT 01 1822 OTHER OFFENSES 01 1491 DRUNKENNESS 01 134 VANDALISM 01 613 SUSPICIOUS OCC 01 353 FRAUD 01 145 LIQUOR LAWS 01 15 LOITERING 01 14 BAD CHECKS 01 1 ASSAULT 01 1483 SEX OFFENSES, FORCIBLE 01 79 DRIVING UNDER THE INFLUENCE 01 93 RECOVERED VEHICLE 01 27 SECONDARY CODES 01 122 ROBBERY 01 644 SUICIDE 01 7 EXTORTION 01 2 KIDNAPPING 01 44 VEHICLE THEFT 01 452 WARRANTS 01 494 BURGLARY 01 301 FORGERY/COUNTERFEITING 01 27 STOLEN PROPERTY 01 61 BRIBERY 01 5 ARSON 01 23 NON-CRIMINAL 01 958 STOLEN PROPERTY 02 44 DRUG/NARCOTIC 02 360 TRESPASS 02 105 BURGLARY 02 315 RECOVERED VEHICLE 02 25 OTHER OFFENSES 02 1211 KIDNAPPING 02 41 SECONDARY CODES 02 117 LIQUOR LAWS 02 10 BRIBERY 02 7 FRAUD 02 75 LARCENY/THEFT 02 1013 SEX OFFENSES, FORCIBLE 02 92 ASSAULT 02 1258 EXTORTION 02 1 DRIVING UNDER THE INFLUENCE 02 80 ARSON 02 28 FORGERY/COUNTERFEITING 02 10 SUICIDE 02 3 MISSING PERSON 02 72 PROSTITUTION 02 325 LOITERING 02 9 WEAPON LAWS 02 119 SUSPICIOUS OCC 02 304 DRUNKENNESS 02 96 DISORDERLY CONDUCT 02 96 NON-CRIMINAL 02 746 VEHICLE THEFT 02 317 WARRANTS 02 408 BAD CHECKS 02 1 ROBBERY 02 782 VANDALISM 02 621 EMBEZZLEMENT 02 2 NON-CRIMINAL 03 470 KIDNAPPING 03 25 FORGERY/COUNTERFEITING 03 13 WARRANTS 03 267 PROSTITUTION 03 179 LARCENY/THEFT 03 648 VANDALISM 03 387 ASSAULT 03 553 SECONDARY CODES 03 72 ROBBERY 03 408 VEHICLE THEFT 03 218 DRUNKENNESS 03 46 DRUG/NARCOTIC 03 210 RECOVERED VEHICLE 03 19 MISSING PERSON 03 47 TRESPASS 03 77 WEAPON LAWS 03 90 STOLEN PROPERTY 03 47 BRIBERY 03 5 SUSPICIOUS OCC 03 221 BURGLARY 03 371 SEX OFFENSES, FORCIBLE 03 68 ARSON 03 36 DISORDERLY CONDUCT 03 53 LIQUOR LAWS 03 5 OTHER OFFENSES 03 843 SUICIDE 03 1 EMBEZZLEMENT 03 1 LOITERING 03 2 FRAUD 03 48 DRIVING UNDER THE INFLUENCE 03 36 NON-CRIMINAL 04 329 VEHICLE THEFT 04 163 DISORDERLY CONDUCT 04 43 SUSPICIOUS OCC 04 140 PROSTITUTION 04 131 TRESPASS 04 64 FORGERY/COUNTERFEITING 04 14 DRIVING UNDER THE INFLUENCE 04 16 STOLEN PROPERTY 04 37 RECOVERED VEHICLE 04 17 SECONDARY CODES 04 37 DRUNKENNESS 04 19 ARSON 04 27 WARRANTS 04 201 MISSING PERSON 04 41 ROBBERY 04 260 FRAUD 04 48 OTHER OFFENSES 04 600 BURGLARY 04 310 SUICIDE 04 1 KIDNAPPING 04 14 LARCENY/THEFT 04 428 VANDALISM 04 237 DRUG/NARCOTIC 04 131 WEAPON LAWS 04 58 LOITERING 04 2 BRIBERY 04 2 ASSAULT 04 410 EMBEZZLEMENT 04 1 SEX OFFENSES, FORCIBLE 04 39 LIQUOR LAWS 04 2 ASSAULT 05 272 WEAPON LAWS 05 29 VANDALISM 05 203 DRIVING UNDER THE INFLUENCE 05 18 STOLEN PROPERTY 05 21 ROBBERY 05 203 MISSING PERSON 05 58 FORGERY/COUNTERFEITING 05 8 BURGLARY 05 255 ARSON 05 26 WARRANTS 05 149 OTHER OFFENSES 05 423 NON-CRIMINAL 05 328 LARCENY/THEFT 05 344 DISORDERLY CONDUCT 05 53 RECOVERED VEHICLE 05 15 SUSPICIOUS OCC 05 107 DRUG/NARCOTIC 05 81 FRAUD 05 31 SEX OFFENSES, FORCIBLE 05 21 DRUNKENNESS 05 8 TRESPASS 05 70 KIDNAPPING 05 14 SECONDARY CODES 05 33 VEHICLE THEFT 05 136 EXTORTION 05 2 PROSTITUTION 05 109 EMBEZZLEMENT 05 1 SUICIDE 05 3 LIQUOR LAWS 05 1 BRIBERY 05 2 GAMBLING 05 1 OTHER OFFENSES 06 650 DRIVING UNDER THE INFLUENCE 06 10 FORGERY/COUNTERFEITING 06 17 WEAPON LAWS 06 39 SEX OFFENSES, FORCIBLE 06 24 TRESPASS 06 101 PORNOGRAPHY/OBSCENE MAT 06 1 EXTORTION 06 1 BURGLARY 06 223 FRAUD

In [0]:
%sql
SELECT Category, Substring(Time,1,2) AS Hour, COUNT(*) AS Count
FROM sf_crime
WHERE PdDistrict ='MISSION'
GROUP BY Category, Hour
ORDER BY Hour

Category Hour Count VANDALISM 00 856 PROSTITUTION 00 706 ROBBERY 00 547 DRUG/NARCOTIC 00 640 BAD CHECKS 00 13 OTHER OFFENSES 00 2703 TRESPASS 00 98 ARSON 00 35 LIQUOR LAWS 00 36 KIDNAPPING 00 40 SECONDARY CODES 00 194 BRIBERY 00 9 VEHICLE THEFT 00 614 WARRANTS 00 666 SEX OFFENSES, NON FORCIBLE 00 3 EXTORTION 00 9 DRIVING UNDER THE INFLUENCE 00 115 GAMBLING 00 6 SUSPICIOUS OCC 00 670 LOITERING 00 23 FORGERY/COUNTERFEITING 00 358 BURGLARY 00 481 WEAPON LAWS 00 188 FRAUD 00 611 RECOVERED VEHICLE 00 40 SUICIDE 00 7 LARCENY/THEFT 00 2553 DISORDERLY CONDUCT 00 188 MISSING PERSON 00 215 ASSAULT 00 1570 DRUNKENNESS 00 155 NON-CRIMINAL 00 1528 EMBEZZLEMENT 00 64 STOLEN PROPERTY 00 71 SEX OFFENSES, FORCIBLE 00 156 TRESPASS 01 71 DISORDERLY CONDUCT 01 129 PROSTITUTION 01 459 WEAPON LAWS 01 157 DRUG/NARCOTIC 01 460 MISSING PERSON 01 114 LARCENY/THEFT 01 1822 OTHER OFFENSES 01 1491 DRUNKENNESS 01 134 VANDALISM 01 613 SUSPICIOUS OCC 01 353 FRAUD 01 145 LIQUOR LAWS 01 15 LOITERING 01 14 BAD CHECKS 01 1 ASSAULT 01 1483 SEX OFFENSES, FORCIBLE 01 79 DRIVING UNDER THE INFLUENCE 01 93 RECOVERED VEHICLE 01 27 SECONDARY CODES 01 122 ROBBERY 01 644 SUICIDE 01 7 EXTORTION 01 2 KIDNAPPING 01 44 VEHICLE THEFT 01 452 WARRANTS 01 494 BURGLARY 01 301 FORGERY/COUNTERFEITING 01 27 STOLEN PROPERTY 01 61 BRIBERY 01 5 ARSON 01 23 NON-CRIMINAL 01 958 STOLEN PROPERTY 02 44 DRUG/NARCOTIC 02 360 TRESPASS 02 105 BURGLARY 02 315 RECOVERED VEHICLE 02 25 OTHER OFFENSES 02 1211 KIDNAPPING 02 41 SECONDARY CODES 02 117 LIQUOR LAWS 02 10 BRIBERY 02 7 FRAUD 02 75 LARCENY/THEFT 02 1013 SEX OFFENSES, FORCIBLE 02 92 ASSAULT 02 1258 EXTORTION 02 1 DRIVING UNDER THE INFLUENCE 02 80 ARSON 02 28 FORGERY/COUNTERFEITING 02 10 SUICIDE 02 3 MISSING PERSON 02 72 PROSTITUTION 02 325 LOITERING 02 9 WEAPON LAWS 02 119 SUSPICIOUS OCC 02 304 DRUNKENNESS 02 96 DISORDERLY CONDUCT 02 96 NON-CRIMINAL 02 746 VEHICLE THEFT 02 317 WARRANTS 02 408 BAD CHECKS 02 1 ROBBERY 02 782 VANDALISM 02 621 EMBEZZLEMENT 02 2 NON-CRIMINAL 03 470 KIDNAPPING 03 25 FORGERY/COUNTERFEITING 03 13 WARRANTS 03 267 PROSTITUTION 03 179 LARCENY/THEFT 03 648 VANDALISM 03 387 ASSAULT 03 553 SECONDARY CODES 03 72 ROBBERY 03 408 VEHICLE THEFT 03 218 DRUNKENNESS 03 46 DRUG/NARCOTIC 03 210 RECOVERED VEHICLE 03 19 MISSING PERSON 03 47 TRESPASS 03 77 WEAPON LAWS 03 90 STOLEN PROPERTY 03 47 BRIBERY 03 5 SUSPICIOUS OCC 03 221 BURGLARY 03 371 SEX OFFENSES, FORCIBLE 03 68 ARSON 03 36 DISORDERLY CONDUCT 03 53 LIQUOR LAWS 03 5 OTHER OFFENSES 03 843 SUICIDE 03 1 EMBEZZLEMENT 03 1 LOITERING 03 2 FRAUD 03 48 DRIVING UNDER THE INFLUENCE 03 36 NON-CRIMINAL 04 329 VEHICLE THEFT 04 163 DISORDERLY CONDUCT 04 43 SUSPICIOUS OCC 04 140 PROSTITUTION 04 131 TRESPASS 04 64 FORGERY/COUNTERFEITING 04 14 DRIVING UNDER THE INFLUENCE 04 16 STOLEN PROPERTY 04 37 RECOVERED VEHICLE 04 17 SECONDARY CODES 04 37 DRUNKENNESS 04 19 ARSON 04 27 WARRANTS 04 201 MISSING PERSON 04 41 ROBBERY 04 260 FRAUD 04 48 OTHER OFFENSES 04 600 BURGLARY 04 310 SUICIDE 04 1 KIDNAPPING 04 14 LARCENY/THEFT 04 428 VANDALISM 04 237 DRUG/NARCOTIC 04 131 WEAPON LAWS 04 58 LOITERING 04 2 BRIBERY 04 2 ASSAULT 04 410 EMBEZZLEMENT 04 1 SEX OFFENSES, FORCIBLE 04 39 LIQUOR LAWS 04 2 ASSAULT 05 272 WEAPON LAWS 05 29 VANDALISM 05 203 DRIVING UNDER THE INFLUENCE 05 18 STOLEN PROPERTY 05 21 ROBBERY 05 203 MISSING PERSON 05 58 FORGERY/COUNTERFEITING 05 8 BURGLARY 05 255 ARSON 05 26 WARRANTS 05 149 OTHER OFFENSES 05 423 NON-CRIMINAL 05 328 LARCENY/THEFT 05 344 DISORDERLY CONDUCT 05 53 RECOVERED VEHICLE 05 15 SUSPICIOUS OCC 05 107 DRUG/NARCOTIC 05 81 FRAUD 05 31 SEX OFFENSES, FORCIBLE 05 21 DRUNKENNESS 05 8 TRESPASS 05 70 KIDNAPPING 05 14 SECONDARY CODES 05 33 VEHICLE THEFT 05 136 EXTORTION 05 2 PROSTITUTION 05 109 EMBEZZLEMENT 05 1 SUICIDE 05 3 LIQUOR LAWS 05 1 BRIBERY 05 2 GAMBLING 05 1 OTHER OFFENSES 06 650 DRIVING UNDER THE INFLUENCE 06 10 FORGERY/COUNTERFEITING 06 17 WEAPON LAWS 06 39 SEX OFFENSES, FORCIBLE 06 24 TRESPASS 06 101 PORNOGRAPHY/OBSCENE MAT 06 1 EXTORTION 06 1 BURGLARY 06 223 FRAUD

In [0]:
%sql
SELECT Category, Substring(Time,1,2) AS Hour, COUNT(*) AS Count
FROM sf_crime
WHERE PdDistrict ='CENTRAL'
GROUP BY Category, Hour
ORDER BY Hour

Category Hour Count DRUG/NARCOTIC 00 263 VANDALISM 00 746 ARSON 00 22 KIDNAPPING 00 26 LIQUOR LAWS 00 12 SECONDARY CODES 00 139 FORGERY/COUNTERFEITING 00 392 WARRANTS 00 412 BRIBERY 00 3 SUICIDE 00 6 DRIVING UNDER THE INFLUENCE 00 78 BAD CHECKS 00 20 EXTORTION 00 13 EMBEZZLEMENT 00 134 SUSPICIOUS OCC 00 538 LOITERING 00 11 GAMBLING 00 1 LARCENY/THEFT 00 2376 BURGLARY 00 425 RECOVERED VEHICLE 00 24 FRAUD 00 660 ASSAULT 00 1106 DISORDERLY CONDUCT 00 80 WEAPON LAWS 00 107 MISSING PERSON 00 157 OTHER OFFENSES 00 2123 STOLEN PROPERTY 00 61 DRUNKENNESS 00 145 NON-CRIMINAL 00 1311 ROBBERY 00 313 SEX OFFENSES, FORCIBLE 00 83 PROSTITUTION 00 169 VEHICLE THEFT 00 376 TRESPASS 00 103 WEAPON LAWS 01 121 MISSING PERSON 01 83 DRUG/NARCOTIC 01 212 LARCENY/THEFT 01 1459 OTHER OFFENSES 01 1190 VANDALISM 01 578 DRUNKENNESS 01 177 SUSPICIOUS OCC 01 278 FRAUD 01 112 SEX OFFENSES, FORCIBLE 01 72 SUICIDE 01 5 LOITERING 01 16 LIQUOR LAWS 01 9 RECOVERED VEHICLE 01 21 EMBEZZLEMENT 01 2 EXTORTION 01 1 DRIVING UNDER THE INFLUENCE 01 62 ASSAULT 01 1260 SECONDARY CODES 01 92 ROBBERY 01 383 VEHICLE THEFT 01 206 KIDNAPPING 01 34 WARRANTS 01 339 BURGLARY 01 313 FORGERY/COUNTERFEITING 01 30 ARSON 01 20 STOLEN PROPERTY 01 53 NON-CRIMINAL 01 869 PROSTITUTION 01 169 TRESPASS 01 59 DISORDERLY CONDUCT 01 62 GAMBLING 02 1 ROBBERY 02 372 TRESPASS 02 72 BURGLARY 02 400 RECOVERED VEHICLE 02 22 KIDNAPPING 02 34 OTHER OFFENSES 02 916 BRIBERY 02 2 SECONDARY CODES 02 74 FRAUD 02 84 LARCENY/THEFT 02 920 SEX OFFENSES, FORCIBLE 02 66 DRIVING UNDER THE INFLUENCE 02 59 ASSAULT 02 1073 ARSON 02 12 PROSTITUTION 02 116 FORGERY/COUNTERFEITING 02 16 MISSING PERSON 02 75 WEAPON LAWS 02 92 SUSPICIOUS OCC 02 269 DRUNKENNESS 02 126 DISORDERLY CONDUCT 02 60 NON-CRIMINAL 02 611 WARRANTS 02 268 VEHICLE THEFT 02 189 VANDALISM 02 567 DRUG/NARCOTIC 02 172 STOLEN PROPERTY 02 35 LIQUOR LAWS 02 15 LOITERING 02 16 EXTORTION 02 4 PORNOGRAPHY/OBSCENE MAT 02 2 PROSTITUTION 03 91 LARCENY/THEFT 03 552 ASSAULT 03 372 VANDALISM 03 386 ROBBERY 03 210 VEHICLE THEFT 03 113 DRUNKENNESS 03 25 DRUG/NARCOTIC 03 102 MISSING PERSON 03 51 RECOVERED VEHICLE 03 13 WEAPON LAWS 03 34 TRESPASS 03 62 STOLEN PROPERTY 03 33 SUSPICIOUS OCC 03 135 BURGLARY 03 374 SEX OFFENSES, FORCIBLE 03 42 ARSON 03 23 OTHER OFFENSES 03 413 FORGERY/COUNTERFEITING 03 13 DRIVING UNDER THE INFLUENCE 03 11 SECONDARY CODES 03 34 DISORDERLY CONDUCT 03 20 LOITERING 03 7 BRIBERY 03 1 SUICIDE 03 1 LIQUOR LAWS 03 1 FRAUD 03 54 KIDNAPPING 03 20 NON-CRIMINAL 03 334 WARRANTS 03 180 NON-CRIMINAL 04 229 SUSPICIOUS OCC 04 102 DISORDERLY CONDUCT 04 21 TRESPASS 04 55 FORGERY/COUNTERFEITING 04 14 STOLEN PROPERTY 04 27 ARSON 04 20 DRUNKENNESS 04 15 SECONDARY CODES 04 33 WARRANTS 04 140 MISSING PERSON 04 33 ROBBERY 04 118 FRAUD 04 31 OTHER OFFENSES 04 252 LARCENY/THEFT 04 393 VEHICLE THEFT 04 93 BURGLARY 04 333 VANDALISM 04 278 DRUG/NARCOTIC 04 61 WEAPON LAWS 04 14 SEX OFFENSES, FORCIBLE 04 12 PROSTITUTION 04 99 RECOVERED VEHICLE 04 10 SUICIDE 04 3 KIDNAPPING 04 7 LOITERING 04 6 DRIVING UNDER THE INFLUENCE 04 5 EMBEZZLEMENT 04 3 LIQUOR LAWS 04 1 BRIBERY 04 1 ASSAULT 04 191 DRIVING UNDER THE INFLUENCE 05 4 ROBBERY 05 94 MISSING PERSON 05 36 STOLEN PROPERTY 05 22 BURGLARY 05 290 ARSON 05 13 OTHER OFFENSES 05 210 WARRANTS 05 66 LARCENY/THEFT 05 431 NON-CRIMINAL 05 239 SUSPICIOUS OCC 05 86 DRUG/NARCOTIC 05 32 EMBEZZLEMENT 05 2 FRAUD 05 29 SECONDARY CODES 05 24 WEAPON LAWS 05 13 TRESPASS 05 37 VEHICLE THEFT 05 88 FORGERY/COUNTERFEITING 05 7 PROSTITUTION 05 92 LOITERING 05 6 DISORDERLY CONDUCT 05 13 GAMBLING 05 2 ASSAULT 05 196 KIDNAPPING 05 11 SUICIDE 05 5 VANDALISM 05 222 DRUNKENNESS 05 4 SEX OFFENSES, FORCIBLE 05 21 RECOVERED VEHICLE 05 5 FORGERY/COUNTERFEITING 06 19 WEAPON LAWS 06 17 LARCENY/THEFT 06 606 ARSON 06 8 TRESPASS 06 75 DRIVING UNDER THE INFLUENCE 06 3 DISORDERLY CONDUCT 06 35 FRAUD 06 46 BURGLARY 06 244 SECONDARY CODES 06 26 SUSPICIOUS OCC 06 110 PROSTITUTION 06 39 KIDNAPPING 06 5 DRUNKENNESS 06 7 DRUG/NARCOTIC 06 48 SEX OFFENSES, FORCIBLE 06 19 VAN

In [0]:
%sql
SELECT Category, Substring(Time,1,2) AS Hour, COUNT(*) AS Count
FROM sf_crime
WHERE PdDistrict ='CENTRAL'
GROUP BY Category, Hour
ORDER BY Hour

Category Hour Count DRUG/NARCOTIC 00 263 VANDALISM 00 746 ARSON 00 22 KIDNAPPING 00 26 LIQUOR LAWS 00 12 SECONDARY CODES 00 139 FORGERY/COUNTERFEITING 00 392 WARRANTS 00 412 BRIBERY 00 3 SUICIDE 00 6 DRIVING UNDER THE INFLUENCE 00 78 BAD CHECKS 00 20 EXTORTION 00 13 EMBEZZLEMENT 00 134 SUSPICIOUS OCC 00 538 LOITERING 00 11 GAMBLING 00 1 LARCENY/THEFT 00 2376 BURGLARY 00 425 RECOVERED VEHICLE 00 24 FRAUD 00 660 ASSAULT 00 1106 DISORDERLY CONDUCT 00 80 WEAPON LAWS 00 107 MISSING PERSON 00 157 OTHER OFFENSES 00 2123 STOLEN PROPERTY 00 61 DRUNKENNESS 00 145 NON-CRIMINAL 00 1311 ROBBERY 00 313 SEX OFFENSES, FORCIBLE 00 83 PROSTITUTION 00 169 VEHICLE THEFT 00 376 TRESPASS 00 103 WEAPON LAWS 01 121 MISSING PERSON 01 83 DRUG/NARCOTIC 01 212 LARCENY/THEFT 01 1459 OTHER OFFENSES 01 1190 VANDALISM 01 578 DRUNKENNESS 01 177 SUSPICIOUS OCC 01 278 FRAUD 01 112 SEX OFFENSES, FORCIBLE 01 72 SUICIDE 01 5 LOITERING 01 16 LIQUOR LAWS 01 9 RECOVERED VEHICLE 01 21 EMBEZZLEMENT 01 2 EXTORTION 01 1 DRIVING UNDER THE INFLUENCE 01 62 ASSAULT 01 1260 SECONDARY CODES 01 92 ROBBERY 01 383 VEHICLE THEFT 01 206 KIDNAPPING 01 34 WARRANTS 01 339 BURGLARY 01 313 FORGERY/COUNTERFEITING 01 30 ARSON 01 20 STOLEN PROPERTY 01 53 NON-CRIMINAL 01 869 PROSTITUTION 01 169 TRESPASS 01 59 DISORDERLY CONDUCT 01 62 GAMBLING 02 1 ROBBERY 02 372 TRESPASS 02 72 BURGLARY 02 400 RECOVERED VEHICLE 02 22 KIDNAPPING 02 34 OTHER OFFENSES 02 916 BRIBERY 02 2 SECONDARY CODES 02 74 FRAUD 02 84 LARCENY/THEFT 02 920 SEX OFFENSES, FORCIBLE 02 66 DRIVING UNDER THE INFLUENCE 02 59 ASSAULT 02 1073 ARSON 02 12 PROSTITUTION 02 116 FORGERY/COUNTERFEITING 02 16 MISSING PERSON 02 75 WEAPON LAWS 02 92 SUSPICIOUS OCC 02 269 DRUNKENNESS 02 126 DISORDERLY CONDUCT 02 60 NON-CRIMINAL 02 611 WARRANTS 02 268 VEHICLE THEFT 02 189 VANDALISM 02 567 DRUG/NARCOTIC 02 172 STOLEN PROPERTY 02 35 LIQUOR LAWS 02 15 LOITERING 02 16 EXTORTION 02 4 PORNOGRAPHY/OBSCENE MAT 02 2 PROSTITUTION 03 91 LARCENY/THEFT 03 552 ASSAULT 03 372 VANDALISM 03 386 ROBBERY 03 210 VEHICLE THEFT 03 113 DRUNKENNESS 03 25 DRUG/NARCOTIC 03 102 MISSING PERSON 03 51 RECOVERED VEHICLE 03 13 WEAPON LAWS 03 34 TRESPASS 03 62 STOLEN PROPERTY 03 33 SUSPICIOUS OCC 03 135 BURGLARY 03 374 SEX OFFENSES, FORCIBLE 03 42 ARSON 03 23 OTHER OFFENSES 03 413 FORGERY/COUNTERFEITING 03 13 DRIVING UNDER THE INFLUENCE 03 11 SECONDARY CODES 03 34 DISORDERLY CONDUCT 03 20 LOITERING 03 7 BRIBERY 03 1 SUICIDE 03 1 LIQUOR LAWS 03 1 FRAUD 03 54 KIDNAPPING 03 20 NON-CRIMINAL 03 334 WARRANTS 03 180 NON-CRIMINAL 04 229 SUSPICIOUS OCC 04 102 DISORDERLY CONDUCT 04 21 TRESPASS 04 55 FORGERY/COUNTERFEITING 04 14 STOLEN PROPERTY 04 27 ARSON 04 20 DRUNKENNESS 04 15 SECONDARY CODES 04 33 WARRANTS 04 140 MISSING PERSON 04 33 ROBBERY 04 118 FRAUD 04 31 OTHER OFFENSES 04 252 LARCENY/THEFT 04 393 VEHICLE THEFT 04 93 BURGLARY 04 333 VANDALISM 04 278 DRUG/NARCOTIC 04 61 WEAPON LAWS 04 14 SEX OFFENSES, FORCIBLE 04 12 PROSTITUTION 04 99 RECOVERED VEHICLE 04 10 SUICIDE 04 3 KIDNAPPING 04 7 LOITERING 04 6 DRIVING UNDER THE INFLUENCE 04 5 EMBEZZLEMENT 04 3 LIQUOR LAWS 04 1 BRIBERY 04 1 ASSAULT 04 191 DRIVING UNDER THE INFLUENCE 05 4 ROBBERY 05 94 MISSING PERSON 05 36 STOLEN PROPERTY 05 22 BURGLARY 05 290 ARSON 05 13 OTHER OFFENSES 05 210 WARRANTS 05 66 LARCENY/THEFT 05 431 NON-CRIMINAL 05 239 SUSPICIOUS OCC 05 86 DRUG/NARCOTIC 05 32 EMBEZZLEMENT 05 2 FRAUD 05 29 SECONDARY CODES 05 24 WEAPON LAWS 05 13 TRESPASS 05 37 VEHICLE THEFT 05 88 FORGERY/COUNTERFEITING 05 7 PROSTITUTION 05 92 LOITERING 05 6 DISORDERLY CONDUCT 05 13 GAMBLING 05 2 ASSAULT 05 196 KIDNAPPING 05 11 SUICIDE 05 5 VANDALISM 05 222 DRUNKENNESS 05 4 SEX OFFENSES, FORCIBLE 05 21 RECOVERED VEHICLE 05 5 FORGERY/COUNTERFEITING 06 19 WEAPON LAWS 06 17 LARCENY/THEFT 06 606 ARSON 06 8 TRESPASS 06 75 DRIVING UNDER THE INFLUENCE 06 3 DISORDERLY CONDUCT 06 35 FRAUD 06 46 BURGLARY 06 244 SECONDARY CODES 06 26 SUSPICIOUS OCC 06 110 PROSTITUTION 06 39 KIDNAPPING 06 5 DRUNKENNESS 06 7 DRUG/NARCOTIC 06 48 SEX OFFENSES, FORCIBLE 06 19 VAN

Step 1: The top-3 danger districts are Southern, Mission, and Northern.
Step 2: In the top-3 danger districts, the crime numbers generally even distributed in each category. At midnight, noon, and around dinner time (16~18 o'clock), the crimes reaches local peaks.
Step 3: By observation and comparison of three districts, it is better to add police during midnight, noon and dinner time. Pay attention to the 'Theft' type of crime in Southern and Central area due to majority of crime type.

#### Q7 question (OLAP)
For different category of crime, find the percentage of resolution. Based on the output, give your hints to adjust the policy.

In [0]:
%python
Num_Res = spark.sql("Select Category, Resolution, count(*) as N_res from sf_crime group by Category, Resolution")
Num_Cat = spark.sql("select Category, count(*) as N_cat from sf_crime group by Category")
Num_Res.createOrReplaceTempView("Num_Res")
Num_Cat.createOrReplaceTempView("Num_Cat")

In [0]:
%sql 
SELECT DISTINCT sf_crime.Category, sf_crime.Resolution, N_res/N_cat AS Percentage
FROM (sf_crime LEFT JOIN Num_Res ON sf_crime.Category=Num_Res.Category and sf_crime.Resolution=Num_Res.Resolution) LEFT JOIN Num_Cat on sf_crime.Category=Num_Cat.Category
WHERE sf_crime.Category='LARCENY/THEFT'
ORDER BY Category, Resolution

Category Resolution Percentage LARCENY/THEFT ARREST, BOOKED 0.052588524504419684 LARCENY/THEFT ARREST, CITED 0.021890266227313143 LARCENY/THEFT COMPLAINANT REFUSES TO PROSECUTE 9.854071865683352E-4 LARCENY/THEFT DISTRICT ATTORNEY REFUSES TO PROSECUTE 9.770385480412154E-4 LARCENY/THEFT EXCEPTIONAL CLEARANCE 9.247345572467179E-4 LARCENY/THEFT LOCATED 9.205502379831582E-5 LARCENY/THEFT NONE 0.9162131910664784 LARCENY/THEFT NOT PROSECUTED 0.0033474554108478475 LARCENY/THEFT PROSECUTED BY OUTSIDE AGENCY 3.2637690255766513E-4 LARCENY/THEFT PROSECUTED FOR LESSER OFFENSE 2.092159631779905E-6 LARCENY/THEFT PSYCHOPATHIC CASE 9.833150269365553E-5 LARCENY/THEFT UNFOUNDED 0.0025545269104032637

In [0]:
%sql 
SELECT DISTINCT sf_crime.Category, sf_crime.Resolution, N_res/N_cat AS Percentage
FROM (sf_crime LEFT JOIN Num_Res ON sf_crime.Category=Num_Res.Category and sf_crime.Resolution=Num_Res.Resolution) LEFT JOIN Num_Cat on sf_crime.Category=Num_Cat.Category
WHERE sf_crime.Category='LARCENY/THEFT'
ORDER BY Category, Resolution

Category Resolution Percentage LARCENY/THEFT ARREST, BOOKED 0.052588524504419684 LARCENY/THEFT ARREST, CITED 0.021890266227313143 LARCENY/THEFT COMPLAINANT REFUSES TO PROSECUTE 9.854071865683352E-4 LARCENY/THEFT DISTRICT ATTORNEY REFUSES TO PROSECUTE 9.770385480412154E-4 LARCENY/THEFT EXCEPTIONAL CLEARANCE 9.247345572467179E-4 LARCENY/THEFT LOCATED 9.205502379831582E-5 LARCENY/THEFT NONE 0.9162131910664784 LARCENY/THEFT NOT PROSECUTED 0.0033474554108478475 LARCENY/THEFT PROSECUTED BY OUTSIDE AGENCY 3.2637690255766513E-4 LARCENY/THEFT PROSECUTED FOR LESSER OFFENSE 2.092159631779905E-6 LARCENY/THEFT PSYCHOPATHIC CASE 9.833150269365553E-5 LARCENY/THEFT UNFOUNDED 0.0025545269104032637

We can conclude that 92% of 'LARCENY/THEFT' type of crimes are not resolved. More police effort is suggested on preventing this type of crimes such as adding more police stuffs in most crimes' area or placing more monitors in allowed lots and plazas.

#### Q8 question: 
Analysis the new columns of the data and find how to use the new columns (e.g., like 'Fire Prevention Districts' etc)

In [0]:
%sql
SELECT `DELETE - Fire Prevention Districts 2 2` as FPD, COUNT(*) AS Count
from sf_crime
GROUP BY FPD
ORDER BY Count DESC

FPD Count 14 232014 8 213731 10 212514 7 206963 2 185688 13 181355 9 166308 15 149331 1 114761 5 114336 11 104394 3 99675 6 65322 12 57092 4 20832 null 5209

In [0]:
%sql
SELECT `DELETE - Fire Prevention Districts 2 2` as FPD, COUNT(*) AS Count
from sf_crime
GROUP BY FPD
ORDER BY Count DESC

FPD Count 14 232014 8 213731 10 212514 7 206963 2 185688 13 181355 9 166308 15 149331 1 114761 5 114336 11 104394 3 99675 6 65322 12 57092 4 20832 null 5209

First, we can see that the fire prevention district 14 has the highest crime rate. Null region and district 4 have obviously much lower crime rate than the other district, which indicates the most safe places.

In [0]:
%sql
SELECT PdDistrict, `DELETE - Fire Prevention Districts 2 2` as FPD, COUNT(*) AS Count
from sf_crime
GROUP BY FPD, PdDistrict
ORDER BY Count DESC

PdDistrict FPD Count BAYVIEW 10 183274 SOUTHERN 14 158393 NORTHERN 13 148413 MISSION 2 128180 MISSION 8 118743 INGLESIDE 9 117488 TARAVAL 1 107272 PARK 15 100134 TENDERLOIN 7 99754 CENTRAL 3 98827 NORTHERN 7 90903 RICHMOND 11 90433 CENTRAL 5 87039 SOUTHERN 8 86435 SOUTHERN 6 65295 TENDERLOIN 14 61148 SOUTHERN 12 56527 INGLESIDE 2 53412 TARAVAL 9 42357 NORTHERN 15 21813 TENDERLOIN 5 19999 MISSION 10 19663 CENTRAL 13 18926 MISSION 15 16274 CENTRAL 4 16031 BAYVIEW 14 12266 RICHMOND 15 11034 RICHMOND 13 10950 SOUTHERN 7 10193 INGLESIDE 10 9352 PARK 11 7741 PARK 1 6982 BAYVIEW 9 6078 MISSION 7 6043 TARAVAL 11 5671 SOUTHERN 4 4791 SOUTHERN 5 4201 PARK 2 3961 SOUTHERN null 3567 TENDERLOIN 8 3464 BAYVIEW 8 3301 NORTHERN 5 3083 TENDERLOIN 13 2474 NORTHERN 8 1787 SOUTHERN 3 790 PARK 13 567 CENTRAL 12 555 INGLESIDE null 513 CENTRAL null 513 NORTHERN 11 362 SOUTHERN 9 358 INGLESIDE 1 311 RICHMOND null 282 BAYVIEW null 152 BAYVIEW 11 151 PARK 14 151 PARK 10 126 TARAVAL null 107 RICHMOND 1 98 BAYVIEW 1 76 SOUTHERN 10 65 MISSION 14 54 SOUTHERN 2 54 BAYVIEW 7 49 TARAVAL 2 39 BAYVIEW 2 35 TENDERLOIN null 34 NORTHERN null 34 TENDERLOIN 15 26 NORTHERN 3 25 TENDERLOIN 3 24 BAYVIEW 13 21 PARK 7 19 SOUTHERN 11 18 TENDERLOIN 10 18 BAYVIEW 15 18 BAYVIEW 6 16 INGLESIDE 15 16 TARAVAL 15 15 BAYVIEW 5 14 CENTRAL 10 13 MISSION 9 13 MISSION 1 13 TENDERLOIN 11 13 PARK 9 12 BAYVIEW 4 10 BAYVIEW 12 10 CENTRAL 6 10 NORTHERN 1 9 BAYVIEW 3 9 NORTHERN 2 6 CENTRAL 11 5 SOUTHERN 13 4 PARK null 4 RICHMOND 10 3 CENTRAL 9 2 RICHMOND 14 2 CENTRAL 7 2 MISSION null 2 RICHMOND 6 1 RICHMOND 2 1 PARK 8 1 SOUTHERN 15 1 NA null 1

In [0]:
%sql
SELECT PdDistrict, `DELETE - Fire Prevention Districts 2 2` as FPD, COUNT(*) AS Count
from sf_crime
GROUP BY FPD, PdDistrict
ORDER BY Count DESC

PdDistrict FPD Count BAYVIEW 10 183274 SOUTHERN 14 158393 NORTHERN 13 148413 MISSION 2 128180 MISSION 8 118743 INGLESIDE 9 117488 TARAVAL 1 107272 PARK 15 100134 TENDERLOIN 7 99754 CENTRAL 3 98827 NORTHERN 7 90903 RICHMOND 11 90433 CENTRAL 5 87039 SOUTHERN 8 86435 SOUTHERN 6 65295 TENDERLOIN 14 61148 SOUTHERN 12 56527 INGLESIDE 2 53412 TARAVAL 9 42357 NORTHERN 15 21813 TENDERLOIN 5 19999 MISSION 10 19663 CENTRAL 13 18926 MISSION 15 16274 CENTRAL 4 16031 BAYVIEW 14 12266 RICHMOND 15 11034 RICHMOND 13 10950 SOUTHERN 7 10193 INGLESIDE 10 9352 PARK 11 7741 PARK 1 6982 BAYVIEW 9 6078 MISSION 7 6043 TARAVAL 11 5671 SOUTHERN 4 4791 SOUTHERN 5 4201 PARK 2 3961 SOUTHERN null 3567 TENDERLOIN 8 3464 BAYVIEW 8 3301 NORTHERN 5 3083 TENDERLOIN 13 2474 NORTHERN 8 1787 SOUTHERN 3 790 PARK 13 567 CENTRAL 12 555 INGLESIDE null 513 CENTRAL null 513 NORTHERN 11 362 SOUTHERN 9 358 INGLESIDE 1 311 RICHMOND null 282 BAYVIEW null 152 BAYVIEW 11 151 PARK 14 151 PARK 10 126 TARAVAL null 107 RICHMOND 1 98 BAYVIEW 1 76 SOUTHERN 10 65 MISSION 14 54 SOUTHERN 2 54 BAYVIEW 7 49 TARAVAL 2 39 BAYVIEW 2 35 TENDERLOIN null 34 NORTHERN null 34 TENDERLOIN 15 26 NORTHERN 3 25 TENDERLOIN 3 24 BAYVIEW 13 21 PARK 7 19 SOUTHERN 11 18 TENDERLOIN 10 18 BAYVIEW 15 18 BAYVIEW 6 16 INGLESIDE 15 16 TARAVAL 15 15 BAYVIEW 5 14 CENTRAL 10 13 MISSION 9 13 MISSION 1 13 TENDERLOIN 11 13 PARK 9 12 BAYVIEW 4 10 BAYVIEW 12 10 CENTRAL 6 10 NORTHERN 1 9 BAYVIEW 3 9 NORTHERN 2 6 CENTRAL 11 5 SOUTHERN 13 4 PARK null 4 RICHMOND 10 3 CENTRAL 9 2 RICHMOND 14 2 CENTRAL 7 2 MISSION null 2 RICHMOND 6 1 RICHMOND 2 1 PARK 8 1 SOUTHERN 15 1 NA null 1

Comparing with police department district, we can see that for the lowest crime FPD region 4 may not be safe because it locates at the central district that ranks No.4 highest crime. The low crime rate in FPD 4 may be because of the relative smaller area. With observation of two plots we can see that FPD 11 that majorly located at RICHMOND may be more safer than other FPD area because it ranks relative low in both plots.

##Conclusion
###This report is based on the crime data in San Francisco area from 2003 to 2018. For the most popular area SF downtown, the peak crime numbers occurred in year 2013. At May 2018, the crime numbers reached the lowest value 3519. 
###By summarizing the crime categories, we found that the top three crimes include 'Larceny/Theft', 'Other offenses' and 'Non-criminal' crimes, among which, 92% of 'LARCENY/THEFT' type of crimes are not resolved. More police effort is suggested on preventing this type of crimes such as adding more police stuffs in most crimes' area or placing more monitors in allowed lots and plazas. The southern district highly overwhelms the other teritory for the number of crime numbers. We suggested more police effort to be placed in this district.
###For the concerned four years (2015~2018) that requested by the customer, the amount of crime numbers showed a decreasing trend. At May 2018, the crime numbers reached the lowest value 3519. It seems that at odd month(like Jan, Mar, May), except 2018, the crime number increased. We may invest insurance company stocks at even month (like Feb, April .etc) because they won't be charged a lot for loss in crimes. On the other hand, we may increase investment in security company stocks at odd month because people may be raised attention for personal and property security.